script/validator.go

```diff
@@ -0,0 +1,198 @@
+}
+
+func (v *Validator) AddError(section, value, message string) {
+	v.Errors = append(v.Errors, ValidationError{
+		Section: section,
+		Value:   value,
+		Message: message,
+	})
+}
+
+func (v *Validator) HasError(section string) bool {
+	for _, err := range v.Errors {
+		if err.Section == section {
+			return true
+		}
+	}
+	return false
+}
+
+func When(condition bool, callback ValidatorCallback) ValidatorCallback {
+	if condition {
+		return callback
+	}
+
+	return func(value string) (bool, string, string) {
+		return true, value, ""
+	}
+}
+
+func ParseInput(value string) (bool, string, string) {
+	if value == "" || value == "_No response_" || value == "None" {
+		return true, "", ""
+	}
+
+	return true, value, ""
+}
+
+func ParsePlainString(value string) (bool, string, string) {
+	// strip all formattig, except for newlines
+	html := blackfriday.Run([]byte(value))
+	doc, err := goquery.NewDocumentFromReader(bytes.NewReader(html))
+	if err != nil {
+		return false, value, err.Error()
+	}
+	value = strings.TrimSpace(doc.Text())
+
+	if urlRegex.MatchString(value) {
+		return false, value, "cannot contain URLs"
+	}
+
+	if emailRegex.MatchString(value) {
+		return false, value, "cannot contain email addresses"
+	}
+
+	if emojiRegex.MatchString(value) {
+		return false, value, "cannot contain emoji characters"
+	}
+
+	return true, value, ""
+}
+
+func ParseAccountUrl(value string) (bool, string, string) {
+	if accountUrlRegex.Match([]byte(value)) {
+		if !strings.HasPrefix(value, "http://") && !strings.HasPrefix(value, "https://") {
+			value = "https://" + value
+		}
+
+		u, err := url.Parse(value)
+		if err != nil {
+			return false, value, err.Error()
+		}
+
+		return true, u.Hostname(), ""
+	} else {
+		return false, value, "is invalid 1Password account URL"
+	}
+}
+
+func ParseCheckbox(value string) (bool, string, string) {
+	value = strings.TrimLeft(strings.ToLower(value), "- ")
+
+	if strings.HasPrefix(value, "[x]") {
+		return true, "true", ""
+	} else if strings.HasPrefix(value, "[]") || strings.HasPrefix(value, "[ ]") {
+		return true, "false", ""
+	}
+
+	return false, value, "could not parse checkbox"
+}
+
+func ParseNumber(value string) (bool, int, string) {
+	cleanedString := ""
+
+	for _, char := range value {
+		if char >= '0' && char <= '9' {
+			cleanedString += string(char)
+		}
+	}
+
+	parsedNumber, err := strconv.Atoi(cleanedString)
+
+	if err != nil {
+		return false, 0, "could not be parsed into a number"
+	}
+
+	return true, parsedNumber, ""
+}
+
+func ParseBool(value string) (bool, bool, string) {
+	parsedBool, err := strconv.ParseBool(value)
+
+	if err != nil {
+		return false, false, "could not be parsed into a boolean"
+	}
+
+	return true, parsedBool, ""
+}
+
+func IsPresent(value string) (bool, string, string) {
+	if value == "" {
+		return false, value, "is empty"
+	}
+
+	return true, value, ""
+}
+
+func IsEmail(value string) (bool, string, string) {
+	if value == "" {
+		return true, value, ""
```

Comment:
Good callout! This and `IsURL` below were initially set up to take `""` as a valid value because the validator executes all validation callbacks; in some instances we didn't want to require a field be filled out (i.e. no `IsPresent`) but still wanted to validate it _if_ it was filled out (i.e. `IsURL`). However you're correct that this could be abused or misunderstood. I've updated the code to make required fields more explicit, and the logic to only validate fields if they are actually set.